cmake/gtest_utils.cmake

```diff
@@ -0,0 +1,76 @@
+enable_testing()
+
+function(k2_add_gtest)
+  cmake_parse_arguments(
+      ARGS                                 # prefix of output variables
+      "NO_MAIN;NO_GMOCK;EXCLUDE_FROM_ALL"  # list of names of the boolean arguments (only defined ones will be true)
+      ""                                   # list of names of mono-valued arguments
+      "SOURCES;DEPENDS"                    # list of names of multi-valued arguments (output variables are lists)
+      ${ARGN}                              # arguments of the function to parse, here we take the all original ones
+  ) # remaining unparsed arguments can be found in ARGS_UNPARSED_ARGUMENTS
+
+  set(dependencies ${ARGS_DEPENDS})
+
+  list(LENGTH ARGS_UNPARSED_ARGUMENTS other_args_size)
+  if(other_args_size EQUAL 1)
+    set(name ${ARGS_UNPARSED_ARGUMENTS})
+  else()
+    message(FATAL_ERROR "too many or not enough unnamed arguments")
+  endif()
+
+  if(ARGS_SOURCES)
+    set(sources ${ARGS_SOURCES})
+  else()
+    message(FATAL_ERROR "you must specify at least one source for the test")
+  endif()
+
+  if(ARGS_EXCLUDE_FROM_ALL)
+    set(EXCLUDE_FROM_ALL "EXCLUDE_FROM_ALL")
+  endif()
+
+  get_filename_component(tgt ${name} LAST_EXT)
+  string(REGEX REPLACE "^[^.]*\\." "" tgt ${name})
+  add_executable(${tgt} ${sources} ${EXCLUDE_FROM_ALL})
+
+  if(ARGS_NO_GMOCK)
+    target_link_libraries(${tgt} PUBLIC ${dependencies} gtest)
+  else()
+    target_link_libraries(${tgt} PUBLIC ${dependencies} gtest gmock)
+  endif()
+
+  set(${CMAKE_PROJECT_NAME}_all_tests "${${CMAKE_PROJECT_NAME}_all_tests};${tgt}" CACHE STRING "" FORCE)
+
+  if(ARGS_NO_MAIN)
+  elseif(ARGS_NO_GMOCK)
+    target_link_libraries(${tgt} PRIVATE gtest_main)
+  else()
+    target_link_libraries(${tgt} PRIVATE gmock_main)
+  endif()
+
+  add_test(NAME ${name} COMMAND $<TARGET_FILE:${tgt}>)
+endfunction()
+
+# directly give the suite name source files,
+function(k2_add_gtests test_set_name)
```

Comment:
Is the following template
https://github.com/danpovey/k2/blob/2a6ba24b4a7845bab757fbf7044b0d6253bcb2d8/k2/csrc/CMakeLists.txt#L29-L41
not good enough that you want to make your own?

I find this cmake script hard to read and understand.

Reply:
I agree it's not quite readable. My motivation was to make this function reusable (and could keep gtest-compile out of `make` process, e.g. only compile and run if `make test`.). As simply put the above codes into a CMake module, it still requires the sources and dep-libs. If you will, please help make this function reusable?